File: tests/ViterbiTest.cpp
/*
 * ViterbiTest.cpp
 *
 *  Created on: 15/09/2017
 *      Author: Lucas Teske
 */

#include "ViterbiTest.h"

#include <random>
#include <iostream>
#include <sathelper.h>

#define FRAMEBYTES 2048
#define FRAMEBITS (FRAMEBYTES * 8)

namespace SatHelper {

    using namespace std;
    unsigned char symbols[8 * 2 * (FRAMEBYTES + 6)];
    unsigned char bits[FRAMEBYTES];
    unsigned char output[FRAMEBYTES];

    bool ViterbiTest::RunTest() {
        cout << "  | Generating Test Data" << endl;
        std::uniform_int_distribution<int> signalDist(0, 255);
        Viterbi27 viterbi(FRAMEBITS+6);

        random_device rd;
        mt19937 e2(rd());

        for (int i=0; i<FRAMEBYTES; i++) {
            bits[i] = signalDist(e2) & 0xFF;
        }

        viterbi.encode(bits, symbols);
        cout << "  | Trying to decode" << endl;

        viterbi.decode(symbols, output);

        int errors = 0;

        for (int i=0; i<FRAMEBYTES; i++) {
            if (output[i] != (0xFF - bits[i])) {
                errors++;
            }
        }

        if (errors > 2) { // First / Last byte can have issues due statistical nature
            cerr << "  | Error count > 2: " << errors << endl;
            return false;
        }

        return true;
    }

} /* namespace SatHelper */

RUNTEST(ViterbiTest)

File: gotests/Viterbi_test.go
package SatHelper

import (
	"testing"
	"github.com/OpenSatelliteProject/libsathelper"
)


const FRAMEBYTES = 2048
const FRAMEBITS = FRAMEBYTES * 8

func TestViterbi(t *testing.T) {
	symbols := make([]byte, 8 * (FRAMEBYTES + 6))
	bits := make([]byte, FRAMEBYTES)
	output := make([]byte, FRAMEBYTES)

	viterbi := SatHelper.NewViterbi27(FRAMEBITS+6)

	t.Log("Creating test array")

	for i := 0; i < FRAMEBYTES; i++ {
		bits[i] = randomByte()
	}

	t.Log("Encoding test array")
	viterbi.Encode(&bits[0], &symbols[0])

	t.Log("Decoding test array")
	viterbi.Decode(&symbols[0], &output[0])

	errorCount := 0

	for i := 0; i <  FRAMEBYTES; i++ {
		if output[i] != (0xFF - bits[i]) {
			errorCount += 1
		}
	}

	if errorCount > 2 {  // First / Last byte can have issues due statistical nature
		t.Errorf("  | Error count > 2: %d", errorCount)
	}
}